lib/src/mocks/http.dart

```diff
@@ -160,17 +203,49 @@ class MockHttp {
   }
 
   void when(Uri uri, RequestHandler handler, {String method}) {
+    // Note: there's really no reason to use `_getUriKey()` here - it strips the
+    // fragment and query from the uri, but neither of those pieces of info are
+    // used anywhere else. The consumer should just be expected to pass in an
+    // exact match here. At the next breaking release, this method and related
+    // ones should be cleaned up & clarified.
     String uriKey = _getUriKey(uri);
-    if (!_handlers.containsKey(uriKey)) {
-      _handlers[uriKey] = {};
+    if (!_requestHandlers.containsKey(uriKey)) {
+      _requestHandlers[uriKey] = {};
     }
     if (method == null) {
-      _handlers[uriKey]['*'] = handler;
+      _requestHandlers[uriKey]['*'] = handler;
     } else {
-      _handlers[uriKey][method.toUpperCase()] = handler;
+      _requestHandlers[uriKey][method.toUpperCase()] = handler;
     }
   }
 
+  void whenPattern(Pattern uriPattern, PatternRequestHandler handler,
+      {String method}) {
+    if (!_patternRequestHandlers.containsKey(uriPattern)) {
+      _patternRequestHandlers[uriPattern] = {};
+    }
+    if (method == null) {
+      _patternRequestHandlers[uriPattern]['*'] = handler;
+    } else {
+      _patternRequestHandlers[uriPattern][method.toUpperCase()] = handler;
+    }
+  }
+
+  void _expect(String method, dynamic uri,
+      {Object failWith,
+      Map<String, String> headers,
+      BaseResponse respondWith}) {
+    if (failWith != null && respondWith != null) {
+      throw new ArgumentError('Use failWith OR respondWith, but not both.');
+    }
+    if (failWith == null && respondWith == null) {
```

Comment:
This one cannot be an else, because I'm doing an `XOR` check here. This method requires that `failWith` is given, or `respondWith` is given, but not both.